src/drivers/net/bnxt/bnxt.c

```diff
@@ -2123,17 +2245,257 @@ static void bnxt_adv_nq_index ( struct bnxt *bp, u16 cnt )
+			pci_read_config_dword ( bp->pdev, reg_addr & mask, &reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_GRC:
+			reg_val = bnxt_er_reg_read ( bp, reg_addr );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR0:
+			reg_val = readl ( bp->bar0 + ( reg_addr & mask ) );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR1:
+			reg_val = readl ( bp->bar1 + ( reg_addr & mask ) );
+			break;
+		default:
+			break;
+	}
+	DBGP ( "read_reg_val bp %p addr %x type %x : reg_val = %x\n", bp, reg_addr, reg_type, reg_val );
+	return reg_val;
+}
+
+void bnxt_rst_reg_val ( struct bnxt *bp, u32 reg_addr, u32 reg_val )
+{
+	u32 mask = ER_QCFG_RESET_REG_ADDR_MASK;
+	u32 reg_type = reg_addr & ER_QCFG_RESET_REG_ADDR_SPACE_MASK;
+
+	switch ( reg_type ) {
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_PCIE_CFG:
+			pci_write_config_dword ( bp->pdev, reg_addr & mask, reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_GRC:
+			bnxt_er_reg_write ( bp, reg_addr, reg_val );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR0:
+			writel ( reg_val, bp->bar0 + ( reg_addr & mask ) );
+			break;
+		case ER_QCFG_FW_HEALTH_REG_ADDR_SPACE_BAR1:
+			writel ( reg_val, bp->bar1 + ( reg_addr & mask ) );
+			break;
+		default:
+			break;
+	}
+}
+
+void bnxt_rst_er_registers ( struct bnxt *bp )
+{
+	u32 delay_time = 0;
+	u8 i;
+
+	for ( i = 0; i < bp->er.reg_array_cnt; i++ ) {
+		bnxt_rst_reg_val ( bp, bp->er.rst_reg[i], bp->er.rst_reg_val[i] );
+
+		delay_time = bp->er.delay_after_rst[i];
+		if ( delay_time ) {
+			udelay ( delay_time * 100000 );
+		}
+	}
+
+}
+
+void bnxt_er_task ( struct bnxt* bp, u8 hb_task )
+{
+	u32 present_hb_cnt;
+	unsigned short pci_command, new_command;
+	u8 i;
+
+	DBGP ( "%s\n", __func__ );
+	if ( bp->er.er_rst_on ) {
+		/* Reset already in progress */
+		udelay ( bp->er.normal_wait_period * 100000 );
+		return;
+	}
+
+	if ( hb_task ) {
+		present_hb_cnt = bnxt_er_get_reg_val ( bp,
+			bp->er.fw_hb_reg,
+			bp->er.fw_hb_reg & ER_QCFG_FW_HB_REG_ADDR_SPACE_MASK,
+			ER_QCFG_FW_HB_REG_ADDR_MASK ) ;
+
+		if ( present_hb_cnt != bp->er.last_fw_hb ) {
+			bp->er.last_fw_hb = present_hb_cnt;
+			return;
+		}
+	}
+
+	/* Heartbeat not incrementing, trigger error recovery */
+	DBGP ( "%s(): Trigger Error Recovery\n", __func__ );
+	bp->er.er_rst_on = 1;
+
+	/* Disable bus master */
+	pci_read_config_word ( bp->pdev, PCI_COMMAND, &pci_command );
+	new_command = pci_command & ~PCI_COMMAND_MASTER;
+	pci_write_config_word ( bp->pdev, PCI_COMMAND, new_command );
+
+	/* Free up resources */
+	bnxt_free_rx_iob ( bp );
+
+	/* wait for firmware to be operational */
+	udelay ( bp->er.rst_min_dsecs * 100000 );
+
+	/* Reconfigure the PCI attributes */
+	pci_write_config_word ( bp->pdev, PCI_COMMAND, pci_command );
+
+	if ( hb_task ) {
+		if ( bp->er.master_pf ) {
+			/* wait for master func wait period */
+			udelay ( bp->er.master_wait_period * 100000 );
+
+			/* Reset register values */
+			bnxt_rst_er_registers ( bp );
```

Comment:
Why does the recovery path not simply reset the NIC?  I'm not understanding why we need a separate "recovery reset" mechanism, given that we are presumably already resetting the NIC to a clean state during driver probe (in `bnxt_init_one()`).  Does the standard reset mechanism not work if the firmware has failed?

If the standard reset mechanism does not recover from a firmware failure, then we should also allow for the possibility that `bnxt_init_one()` will need to perform recovery.  Either way: there should be a single code path for "get the NIC into a clean and guaranteed working state", and this single code path should be used by both `bnxt_init_one()` and `bnxt_er_task()`.

Reply:
Error recovery cannot be handled solely by the driver, as it is not capable of independently resetting the NIC.  If the firmware has failed, part of the error recovery process is to restart the firmware so that it is in a healthy state.  The adapter error recovery feature is intended to work without needing to re-probe/re-open the driver.  This is the main reason why there is a separate recovery path.  The standard probe/open path assumes the firmware is healthy and ready, while the recovery path is meant to restore the NIC to a usable state once the firmware has completed its recovery.

Reply:
If the firmware fails prior to iPXE execution, and it has not recovered prior to the driver initialization, the driver would fail to load.  As part of the normal driver initialization, it will query the firmware on error recovery capabilities as well as retrieve error recovery related parameters needed during the "reset" path, if applicable.  The driver must be able to go through a "normal" probe/open at least once in order for it be "error recovery aware".  There are no hardcoded error recovery registers on the part of the driver, as this information is retrieved from the firmware

Reply:
OK, I understand.  This seems like a design flaw in the recovery mechanism: if there's no way for a freshly loaded iPXE or Linux kernel driver (with no saved state) to perform firmware recovery, then that creates a situation in which the only way to proceed is a hard reboot.  But if this hardware flaw does exist then there's nothing we can do about it, so I accept the rationale of keeping recovery separate from normal reset.